### PR TITLE
[RLlib] Shared encoder MARL unittest and example

### PR DESCRIPTION
**Pull request by kouroshHakha**
## Why are these changes needed?
This PR shows how one can create a multi-agent RLModule with shared encoders across two modules. When you train one agent the gradient propragated thru the encoder will get the encoder that lives inside other modules updated. This was not possible in the old stack of APIs. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kouroshHakha** [comment]
@ArturNiederfahrenhorst There you go. You wanted to see this :) ?

**Comment by ArturNiederfahrenhorst** [comment]
Awesome! Yeah I believe this is an important feature. Not only because it's powerful but because I've seen people ask for this or related stuff on discourse. 🙇🏻‍♂️

File: rllib/core/testing/tf/bc_module.py [hunk 1]
```diff
@@ -100,3 +101,78 @@ def from_model_config(
...
+        )
+
+    def input_specs_inference(self):
+        return self._common_input_spec()
+
+    def input_specs_exploration(self):
+        return self._common_input_spec()
+
+    def input_specs_train(self):
+        return self._common_input_spec()
```
Comment (gjoliver) [inline rllib/core/testing/tf/bc_module.py:130]: dude, can these be the defaults? like all 3 input_specs_xxx() functions can _common_input_spec() in the base class, so we don't have to write these all the time?

File: rllib/core/testing/tf/bc_module.py [hunk 2]
```diff
@@ -100,3 +101,78 @@ def from_model_config(
...
+        return [("obs", "global"), ("obs", "local")]
+
+    def _forward_inference(self, batch):
+        return self._common_forward(batch)
+
+    def _forward_exploration(self, batch):
+        return self._common_forward(batch)
+
+    def _forward_train(self, batch):
+        return self._common_forward(batch)
```
Comment (gjoliver) [inline rllib/core/testing/tf/bc_module.py:142]: these too.

File: rllib/core/testing/tf/bc_module.py [hunk 3]
```diff
@@ -100,3 +101,78 @@ def from_model_config(
...
+
+    def _common_forward(self, batch):
+        obs = batch["obs"]
+        breakpoint()
```
Comment (gjoliver) [inline rllib/core/testing/tf/bc_module.py:146]: no breakpoint man

File: rllib/core/testing/torch/bc_module.py [hunk 4]
```diff
@@ -85,3 +86,89 @@ def from_model_config(
...
+    """An example of an RLModule that uses an encoder shared with other things.
+
+    For example, we could consider a multi-agent case where for inference each agent
+    needs to know the global state of the environment, as well as the local state of
+    itself. For better representation learning we would like to share the encoder
+    across all the modules. So this module simply accepts the encoder object as its
+    input argument and uses it to encode the global state. The local state is passed
+    through as is. The policy head is then a simple MLP that takes the concatenation of
+    the global and local state as input and outputs the action logits.
+
```
Comment (gjoliver) [inline rllib/core/testing/torch/bc_module.py:101]: very cool! but how does gradients flow through the shared encoder? it will get some gradients from each single agents modules' backward() right?

File: rllib/core/testing/tf/bc_module.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:130]: The default is different module by module, right? 
For example here, we have global and local under obs. How can this go to base class? You can leave the specs empty if you don't feel like spec checking :-))

Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:130]: The default is different module by module, right? 
For example here, we have global and local under obs. How can this go to base class? You can leave the specs empty if you don't feel like spec checking :-))

Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:130]: The default is different module by module, right? 
For example here, we have global and local under obs. How can this go to base class? You can leave the specs empty if you don't feel like spec checking :-))

Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:130]: The default is different module by module, right? 
For example here, we have global and local under obs. How can this go to base class? You can leave the specs empty if you don't feel like spec checking :-))

File: rllib/core/testing/tf/bc_module.py (hunk 3 above)
Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:146]: Oh shit

Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:146]: Oh shit

Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:146]: Oh shit

Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:146]: Oh shit

File: rllib/core/testing/tf/bc_module.py (hunk 1 above)
Comment (gjoliver) [reply rllib/core/testing/tf/bc_module.py:130]: oh sorry, what I meant was you have
```
def input_specs_train(self):
        return self._common_input_spec()
```
in the base class. for each different module, people then just override ``_common_input_spec()`` once.
the pattern of all three input_specs_xxx sharing the same spec should be common?

File: rllib/core/testing/tf/bc_module.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:142]: forward_train() in most cases is not the same as forward_inference or exploration. So I want the user to explicitly implement them. If you look at pytorch_lightning they also have 3 separate functions, even tho they will all need to call forward method of the underlying module. Being explicit here is intentional. For specs I agree. Cuz for 99% of use-cases the main key that should exist is `OBS` at the very least.

File: rllib/core/testing/tf/bc_module.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/core/testing/tf/bc_module.py:130]: I think for specs it makes sense.

File: rllib/core/testing/torch/bc_module.py (hunk 4 above)
Comment (kouroshHakha) [reply rllib/core/testing/torch/bc_module.py:101]: yes, basically the enablement comes from the fact that you have one memory location for encoder, but the encoder is used in several modules. So if the outputs of encoder is used across several modules, it's gradients would be computed w.r.t. all module losses.

File: rllib/core/testing/tf/bc_module.py (hunk 2 above)
Comment (gjoliver) [reply rllib/core/testing/tf/bc_module.py:137]: I see. that makes sense.

File: rllib/core/testing/torch/bc_module.py (hunk 4 above)
Comment (gjoliver) [reply rllib/core/testing/torch/bc_module.py:101]: cool!

**Review by gjoliver — APPROVED** [review]

**Comment by gjoliver** [comment]
two step qmix now straight fails 😓
